

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>VCFclass - An interface for interacting with VCF files &mdash; VCFClass 0.1.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Typical Workflow" href="usage.html" />
    <link rel="prev" title="VCFclass: VCF interface for python" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> VCFClass
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">VCFclass - An interface for interacting with VCF files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-VCFclass">API</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Typical Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installing VCFclass</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">VCFClass</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>VCFclass - An interface for interacting with VCF files</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="vcfclass-an-interface-for-interacting-with-vcf-files">
<h1>VCFclass - An interface for interacting with VCF files<a class="headerlink" href="#vcfclass-an-interface-for-interacting-with-vcf-files" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>VCFclass is a python module that allows for better integration of
within-sample variation data stored in VCF files and common bioinformatics
tools (Pandas, numpy, etc.) VCFclass is designed to handle the small VCF
files typical of viral sequencing data.</p>
<p>This page provides a quick introduction in using VCFclass followed by the
API. See <a class="reference internal" href="usage.html#usage"><span class="std std-ref">Typical Workflow</span></a> for walkthroughs of typical usage cases.</p>
<p>To use the module to read a file in BAM format, create a
<a class="reference internal" href="#VCFclass.VCF" title="VCFclass.VCF"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCF</span></code></a> object:: python</p>
<blockquote>
<div><p>import VCFclass
my_vcf = VCFclass.VCF(“my_vcf.vcf”)</p>
</div></blockquote>
<p>Once a file is opened you can directly iterate over mutation calls
in the VCF file:: python</p>
<blockquote>
<div><dl class="simple">
<dt>for mut in my_vcf:</dt><dd><p>do_something</p>
</dd>
</dl>
</div></blockquote>
<p>Each iteration returns a <a class="reference internal" href="#VCFclass.MutCall" title="VCFclass.MutCall"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutCall</span></code></a> object.</p>
<p>An alternative way of accessing the data in a VCF file is by
calling <a class="reference internal" href="#VCFclass.VCF.to_dataframe" title="VCFclass.VCF.to_dataframe"><code class="xref py py-meth docutils literal notranslate"><span class="pre">to_dataframe()</span></code></a> method. This function
returns a dataframe containing one row per sample per mutation.
Each row contains information about the mutation call (position,
ref, alt, etc.) along with information specific to each sample
(depth, ref_depth, alt_depth, etc.) This information is particularly
easy to use with downstream tools (e.g., Seaborn, or Pandas’ groupby
functions.)</p>
<p>More detailed usage instructions, including information about how to
average VCFs that are technical replicates and to apply a read position
SNP filter, are at <a class="reference internal" href="usage.html#usage"><span class="std std-ref">Typical Workflow</span></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Coordinates in VCFclass are always 0-based (following the python
convention). Text VCF files, SAM files, and GTF files use 1-based
coordinates. VCFclass converts positions from 1-based coordinates
to 0-based upon import, and converts positions back to 1-based
coordinates upon export to text VCF files. Pandas Dataframes created
using the <code class="xref py py-meth docutils literal notranslate"><span class="pre">to_dataframe()</span></code> method will have 0-based
coordinates.</p>
<p>In general, VCFclass’ design philosophy is to work in 0-based coordinates
at all times unless a format specifies 1-based coordinates.</p>
</div>
</div>
<div class="section" id="module-VCFclass">
<span id="api"></span><h2>API<a class="headerlink" href="#module-VCFclass" title="Permalink to this headline">¶</a></h2>
<p>An <a class="reference internal" href="#VCFclass.VCF" title="VCFclass.VCF"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCF</span></code></a> represents a whole VCF file.</p>
<dl class="py class">
<dt class="sig sig-object py" id="VCFclass.VCF">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">VCFclass.</span></span><span class="sig-name descname"><span class="pre">VCF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">location</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refFile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtfFile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bamfiles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.VCF" title="Permalink to this definition">¶</a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="VCFclass.VCF.averageSamples">
<span class="sig-name descname"><span class="pre">averageSamples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sampleA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sampleB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">newSampleName</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.VCF.averageSamples" title="Permalink to this definition">¶</a></dt>
<dd><p>take intersection of mutations and average their coverages/frequencies etc.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="VCFclass.VCF.averageWithVCF">
<span class="sig-name descname"><span class="pre">averageWithVCF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">otherVCF</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">newSampleName</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.VCF.averageWithVCF" title="Permalink to this definition">¶</a></dt>
<dd><p><em>assumes one sample per vcf</em>
take intersection of mutations and average their coverages/frequencies etc.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="VCFclass.VCF.removemut">
<span class="sig-name descname"><span class="pre">removemut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.VCF.removemut" title="Permalink to this definition">¶</a></dt>
<dd><p>muts are stored in self.mutations and self._hashedmuts.
this removes all muts w/ chrom and pos from both lists.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="VCFclass.VCF.renameSample">
<span class="sig-name descname"><span class="pre">renameSample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">origname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">newName</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.VCF.renameSample" title="Permalink to this definition">¶</a></dt>
<dd><p>update name of sample origname in both list of samples (self.samples)
and in all mutations</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="VCFclass.VCF.to_dataframe">
<span class="sig-name descname"><span class="pre">to_dataframe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.VCF.to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>export VCF as tidy dataframe</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="VCFclass.VCF.to_dict">
<span class="sig-name descname"><span class="pre">to_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.VCF.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>returns list of dictionaries. Each dictionary is ‘property’:value for each snp in each sample.</p>
</dd></dl>

</dd></dl>

<p>An <a class="reference internal" href="#VCFclass.MutCall" title="VCFclass.MutCall"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutCall</span></code></a> represents one row of a VCF file.</p>
<dl class="py class">
<dt class="sig sig-object py" id="VCFclass.MutCall">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">VCFclass.</span></span><span class="sig-name descname"><span class="pre">MutCall</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.MutCall" title="Permalink to this definition">¶</a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="VCFclass.MutCall.apply_pos_filter">
<span class="sig-name descname"><span class="pre">apply_pos_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">removeFails</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">snp_frequency_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_read_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_base_quality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mapping_quality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'snp_filter_log.log'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.MutCall.apply_pos_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>given pysam alignment object and relevent sample, determine whether average
in-read positions for ref and alt alleles come from different distributions.
If so, it sets that sample’s minority allele read ct to 0 (since these are presumed to be alignment errors).
If that means the ref allele is 100%, the sample is removed from the mutation.
If that means the alt allele is 100%, the sample is retained with updated statistics.</p>
</dd></dl>

</dd></dl>

<p>An <a class="reference internal" href="#VCFclass.SampleMut" title="VCFclass.SampleMut"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleMut</span></code></a> represents information about a
mutation that is specific to an individual sample.</p>
<dl class="py class">
<dt class="sig sig-object py" id="VCFclass.SampleMut">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">VCFclass.</span></span><span class="sig-name descname"><span class="pre">SampleMut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samplename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formatinfo</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.SampleMut" title="Permalink to this definition">¶</a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="VCFclass.SampleMut.zeroOut">
<span class="sig-name descname"><span class="pre">zeroOut</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.SampleMut.zeroOut" title="Permalink to this definition">¶</a></dt>
<dd><p>If a sample fails a filter or otherwise needs to be deleted in only one sample,
it’s zero-ed out so that it’s blank. Importantly, deleting a sample from one mutation
is different than deleting a sample from the whole vcf.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="VCFclass.Header">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">VCFclass.</span></span><span class="sig-name descname"><span class="pre">Header</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">headertext</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#VCFclass.Header" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="usage.html" class="btn btn-neutral float-right" title="Typical Workflow" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="VCFclass: VCF interface for python" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2021, Joseph Lalli

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>